modified URL openers so they do not check SSL certificates

arucard21 opened the pull request:
This should be the PR that replaces #242. 

Maybe you have a reason for it, but wouldn't it be better if the junalmeida/torrentProviders branch wasn't behind the junalmeida/torrent_1080_subtitles branch? It seems to me like the "development" branch (torrentProviders) should always be ahead of, or even with the stable branch (torrent_1080_subtitles). Anyway, just a suggestion. The earlier PR really did turn into a mess, which was my fault. I hope this one is better, I also included a minor fix in helpers.py.

VeNoMouS commented:
I disagree with pushpullet moving away from requests lib to use urllib2, urllib2 has many many problems with it, I've slowly migrated the providers off urllib2 and you're trying to put it back... 

What are you're reasons your trying to force back urllib2?

VeNoMouS commented:
8e8008f why are you  are you ONLY pulling the first element 
``` auth = _authToken(None,host,username,password)[0] ```

![selection_240](https://cloud.githubusercontent.com/assets/597545/6606918/01cc88a0-c8a0-11e4-94a0-ef15aec6c025.png)

VeNoMouS commented:
Overall I have no problem with the change, there's obviously alot of changes but its all name space changes and indentation with some additional sanity checks added that were missed, Hope you updated all those geturl helpers :P, thank you for the work, The only concern I have is mentioned above with moving away from requests for pushbullet...

VeNoMouS commented:
Hey is it possible to implement an option to by pass ignoring the SSL verify like sickrage have done?

https://github.com/SiCKRAGETV/SickRage/search?utf8=%E2%9C%93&q=sickbeard.TORRENT_VERIFY_CERT

arucard21 commented:
This is basically a port of the fix I created for the midgetspy fork of sickbeard. You can find more detailed information about it in https://github.com/midgetspy/Sick-Beard/pull/912

Basically, I updated the getURL() methods and tried to get all sickbeard code to use those getURL() methods when they want to access a URL. This should allow a more uniform behaviour within sickbeard, but the reason for this PR was actually to fix a problem with HTTPS connections on python 2.7.9+ (you can read more about that in the midgetspy PR)

There's already an optional "validate" parameter to the getURL() methods that is currently set to "False" by default, but you can use a global variable as default as well. You'd only have to change it in that one place. I don't really know how to create a user-configurable global parameter, so I didn't. 

As for the replacing the requests in pushbullet with urllib2. This was done so it could use the getURL() methods for the sake of consistency. If you really want to move away from urllib2 and move to requests, now you only have to change it in the getURL() methods and it would be applied everywhere in sickbeard. I don't really see the need to change to requests though.

The reason I used urllib2 is because urllib and urllib2 were very widely used in sickbeard, so I actually didn't really see the need for an alternative. Some places still used urllib to connect to a URL, so these could be updated to urllib2. But the biggest reason for using urllib2, which is also the reason I think it should not be replaced by requests, is that urllib2 is a core library in python and doesn't require an additional library. If we needed to do some more intricate work with URL's, then it might be useful to use something like requests. But for this simple functionality, it just seems like an unnecessary abstraction which would cause unnecessary overhead.

@VeNoMouS I'm not sure what problems you have with urllib2. Maybe some of them were actually related to the now-deprecated urlopen() methods in urllib (not urllib2)? It seems to work just fine for me and if the complexity of urllib2's usage is the problem, it's now wrapped into the getURL() methods so it shouldn't be that big of a problem. I hope I managed to answer all your questions.

VeNoMouS commented:
@arucard21 There are a couple reasons I have discontinued use of urllib2 in my projects 
a) its not in python3
b) requests (which we have included in lib dir) is a wrapper for urllib3
c) urllib3 benchmarking takes half the time urllib2 does
d) simplicity , you can do a get/post request with half the code required for urllib2
e) requests automatically deflates gzip, where as you have to do it yourself with urllib2

I could go on, but urllib2 is a dying module imho

arucard21 commented:
@VeNoMouS 
I understand what you're saying and I agree that the requests module has merit but I wouldn't say that urllib2 is a dying module. The points you mention don't seem entirely accurate or relevant from what I understand (but I'm no python expert so feel free to correct me).

a) urllib2 is not in python3 but its methods can be automatically converted to python3. The following is from the [urllib2 documentation](https://docs.python.org/2/library/urllib2.html):
`The urllib2 module has been split across several modules in Python 3 named urllib.request and urllib.error. The 2to3 tool will automatically adapt imports when converting your sources to Python 3.`
Additionally, neither requests nor urllib3 is in python3 since they are third-party libraries so this point doesn't seem relevant.

b) this is true, i assume this is relevant because of the next point (otherwise I'm not sure why it would be relevant)

c) I read this on the [urllib3 Github site](https://github.com/shazow/urllib3) as well, but the performance improvement comes from connection re-use/pooling. The test runs twice as fast when trying to access 15 different URL from the same host. I don't think sickbeard does this very often (or at all) so we would not see any significant performance improvement there. The "file post" part of urllib3 might be useful. We use a MultiPartPostHandler to support (large) file uploads, but i admit that it is not that simple.

d) I do agree with this, it might be the biggest benefit.

e) I actually created the readURLFileLike() method for this, so it can be automatically unzipped or deflated here too.

I do like the requests module as I've used it in other projects before, but I think it's better to use built-in libraries as much as possible. This improves the portability of the code and reduces the amount of maintenance that's required. But I'm not sure if the benefits outweigh the drawbacks. I think the main reason to use requests would be to have simpler code in sickbeard at the cost of portability and possibly maintainance (which could be offset by the simpler code we'd have in sickbeard)

The way the requests module is currently added to the lib folder in sickbeard seems particularly difficult to maintain though, using a copy of the code instead of just having that package as a dependency. I'm not sure if there are technical reasons for this, but this means that we need to maintain the requests code in sickbeard as well and we have to update the code with newer code from upstream manually (I think). This means that the code in this lib folder could easily become outdated (compared to the latest upstream) and could be difficult to update, especially if changes were made in the sickbeard version of the code. This also affects the portability of the code as the libs need to be taken into account as well.

Lastly, while I think this is a very interesting topic (as indicated by the massive amount of text here :P ), I think the choice to use urllib2 or requests is not really relevant for this PR. If you want the pushbullet notifier to keep using requests, maybe you can submit it again (copied from an older version or something) as a separate PR. Even with requests though, you'd need to change it so it doesn't verify certificates by default (since the change in behaviour is caused by a change in the ssl module in python which is also used by urllib3 and therefore also by requests).

arucard21 commented:
I've fixed some regressions I found while running this code as my main sickbeard installation. Aside from these now-fixed problems, I haven't run into any big problems yet.

arucard21 commented:
Closed because no longer relevant
